File: internal/analyzers/analyze/schema_registry.go
package analyze

// FieldMeta describes a single field in an analyzer's output schema.
type FieldMeta struct {
	Type        string `json:"type"                  yaml:"type"`
	Grain       string `json:"grain,omitempty"       yaml:"grain,omitempty"`
	Description string `json:"description,omitempty" yaml:"description,omitempty"`
}

// AnalyzerSchema maps output field names to their metadata.
type AnalyzerSchema map[string]FieldMeta

// SchemaForAnalyzer returns the output schema for the given analyzer ID,
// or nil if the analyzer is not registered.
func SchemaForAnalyzer(analyzerID string) AnalyzerSchema {
	schema, ok := analyzerSchemas[analyzerID]
	if !ok {
		return nil
	}

	return schema
}

// analyzerSchemas is the static registry of output schemas for all analyzers.
var analyzerSchemas = map[string]AnalyzerSchema{
	"static/complexity": {
		"function_complexity": {Type: "list", Grain: "function", Description: "Per-function cyclomatic and cognitive complexity"},
		"distribution":        {Type: "aggregate", Description: "Complexity distribution (simple/moderate/complex)"},
		"high_risk_functions": {Type: "risk", Grain: "function", Description: "Functions exceeding complexity thresholds"},
		"aggregate":           {Type: "aggregate", Description: "Summary statistics"},
	},
	"static/halstead": {
		"function_halstead":     {Type: "list", Grain: "function", Description: "Per-function Halstead volume, effort, and bugs"},
		"distribution":          {Type: "aggregate", Description: "Effort distribution (low/medium/high/very_high)"},
		"high_effort_functions": {Type: "risk", Grain: "function", Description: "Functions with high Halstead effort"},
		"aggregate":             {Type: "aggregate", Description: "Summary statistics"},
	},
	"static/cohesion": {
		"function_cohesion":      {Type: "list", Grain: "function", Description: "Per-function LCOM cohesion score"},
		"distribution":           {Type: "aggregate", Description: "Cohesion distribution"},
		"low_cohesion_functions": {Type: "risk", Grain: "function", Description: "Functions with poor cohesion"},
		"aggregate":              {Type: "aggregate", Description: "Summary statistics"},
	},
	"static/comments": {
		"comment_quality":        {Type: "list", Grain: "comment", Description: "Per-comment quality assessment"},
		"function_documentation": {Type: "list", Grain: "function", Description: "Per-function documentation status"},
		"undocumented_functions": {Type: "risk", Grain: "function", Description: "Functions lacking documentation"},
		"aggregate":              {Type: "aggregate", Description: "Summary statistics"},
	},
	"static/clones": {
		"clone_pairs":             {Type: "list", Grain: "pair", Description: "Detected clone pairs with similarity"},
		"clone_type_distribution": {Type: "aggregate", Description: "Clone type breakdown (Type-1/2/3)"},
		"total_functions":         {Type: "scalar", Description: "Total functions analyzed"},
		"total_clone_pairs":       {Type: "scalar", Description: "Total clone pairs (uncapped)"},
		"clone_ratio":             {Type: "scalar", Description: "Fraction of functions involved in duplication"},
	},
	"static/imports": {
		"import_list":  {Type: "list", Grain: "import", Description: "All import statements"},
		"dependencies": {Type: "list", Grain: "dependency", Description: "External dependencies with risk"},
		"categories":   {Type: "aggregate", Description: "Import category breakdown"},
		"aggregate":    {Type: "aggregate", Description: "Summary statistics"},
	},
	"static/composition": {
		"breakdown":   {Type: "aggregate", Description: "File count per category"},
		"percentages": {Type: "aggregate", Description: "Percentage per category"},
		"total_files": {Type: "scalar", Description: "Total files analyzed"},
	},
	"history/sentiment": {
		"time_series":           {Type: "time_series", Grain: "tick", Description: "Per-tick sentiment scores"},
		"trend":                 {Type: "aggregate", Description: "Sentiment trend direction"},
		"low_sentiment_periods": {Type: "risk", Grain: "tick", Description: "Ticks with negative sentiment"},
		"aggregate":             {Type: "aggregate", Description: "Summary statistics"},
	},
	"history/anomaly": {
		"anomalies":   {Type: "list", Grain: "tick", Description: "Detected anomalous ticks"},
		"time_series": {Type: "time_series", Grain: "tick", Description: "Per-tick anomaly metrics and z-scores"},
		"aggregate":   {Type: "aggregate", Description: "Summary statistics"},
	},
	"history/devs": {
		"developers": {Type: "list", Grain: "developer", Description: "Per-developer contribution statistics"},
		"languages":  {Type: "list", Grain: "language", Description: "Per-language contribution breakdown"},
		"busfactor":  {Type: "list", Grain: "language", Description: "Bus factor per language"},
		"activity":   {Type: "time_series", Grain: "tick", Description: "Per-tick commit activity by developer"},
		"churn":      {Type: "time_series", Grain: "tick", Description: "Per-tick lines added/removed"},
		"aggregate":  {Type: "aggregate", Description: "Summary statistics"},
	},
	"history/file-history": {
		"file_churn":        {Type: "list", Grain: "file", Description: "Per-file change frequency and contributors"},
		"file_contributors": {Type: "list", Grain: "file", Description: "Per-file contributor breakdown"},
		"hotspots":          {Type: "risk", Grain: "file", Description: "High-churn files"},
		"composition":       {Type: "aggregate", Description: "File type composition"},
		"composition_ts":    {Type: "time_series", Grain: "tick", Description: "File composition over time"},
		"aggregate":         {Type: "aggregate", Description: "Summary statistics"},
	},
	"history/couples": {
		"file_coupling":      {Type: "list", Grain: "pair", Description: "Co-changed file pairs"},
		"developer_coupling": {Type: "list", Grain: "pair", Description: "Developer collaboration pairs"},
		"file_ownership":     {Type: "list", Grain: "file", Description: "Per-file ownership"},
		"aggregate":          {Type: "aggregate", Description: "Summary statistics"},
	},
	"history/shotness": {
		"node_hotness":  {Type: "list", Grain: "node", Description: "AST node change frequency"},
		"node_coupling": {Type: "list", Grain: "pair", Description: "Co-changed AST node pairs"},
		"hotspot_nodes": {Type: "risk", Grain: "node", Description: "Frequently changed nodes"},
		"aggregate":     {Type: "aggregate", Description: "Summary statistics"},
	},
	"history/burndown": {
		"global_survival":    {Type: "time_series", Grain: "sample", Description: "Global code survival curve"},
		"file_survival":      {Type: "list", Grain: "file", Description: "Per-file survival data"},
		"developer_survival": {Type: "list", Grain: "developer", Description: "Per-developer survival data"},
		"aggregate":          {Type: "aggregate", Description: "Summary statistics"},
	},
	"history/quality": {
		"time_series": {Type: "time_series", Grain: "tick", Description: "Per-tick code quality metrics"},
		"aggregate":   {Type: "aggregate", Description: "Summary statistics"},
	},
	"history/imports": {
		"import_list":  {Type: "list", Grain: "import", Description: "Import statements (requires UAST mode)"},
		"dependencies": {Type: "list", Grain: "dependency", Description: "Dependencies (requires UAST mode)"},
		"categories":   {Type: "aggregate", Description: "Import category breakdown"},
		"aggregate":    {Type: "aggregate", Description: "Summary statistics"},
	},
	"history/typos": {
		"typos": {Type: "list", Grain: "identifier", Description: "Detected identifier typos (requires UAST mode)"},
	},
}

File: rust/crates/cf-analyze/src/schema_registry.rs
//! Static registry of analyzer output schemas.
//!
//! Port of `internal/analyzers/analyze/schema_registry.go`. The schema is
//! attached to [`crate::conversion::AnalyzerResult`] (`schema,omitempty`) and
//! therefore serialized, so [`FieldMeta`]'s field order (`type`, `grain`,
//! `description`) and omitempty are preserved, and the registry contents match
//! the Go table verbatim.

use std::collections::BTreeMap;
use std::sync::OnceLock;

/// Describes a single field in an analyzer's output schema.
///
/// Mirrors `FieldMeta` (schema_registry.go:4). Serialized order is `type`,
/// `grain` (omitempty), `description` (omitempty).
#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct FieldMeta {
    /// Field kind (`type`): `list`, `aggregate`, `risk`, `scalar`,
    /// `time_series`, …
    pub r#type: String,
    /// Optional data grain (`grain,omitempty`): `function`, `file`, `tick`, …
    pub grain: String,
    /// Optional human description (`description,omitempty`).
    pub description: String,
}

impl FieldMeta {
    fn new(r#type: &str, grain: &str, description: &str) -> Self {
        Self {
            r#type: r#type.to_string(),
            grain: grain.to_string(),
            description: description.to_string(),
        }
    }
}

/// Maps output field names to their metadata.
///
/// Mirrors `AnalyzerSchema = map[string]FieldMeta` (schema_registry.go:11). A
/// [`BTreeMap`] gives the byte-sorted key order the encoder would apply to a Go
/// `map[string]…` anyway, keeping serialization byte-identical.
pub type AnalyzerSchema = BTreeMap<String, FieldMeta>;

/// Returns the output schema for `analyzer_id`, or `None` if unregistered.
///
/// Mirrors `SchemaForAnalyzer` (schema_registry.go:15).
#[must_use]
pub fn schema_for_analyzer(analyzer_id: &str) -> Option<AnalyzerSchema> {
    registry().get(analyzer_id).cloned()
}

fn registry() -> &'static BTreeMap<String, AnalyzerSchema> {
    static REGISTRY: OnceLock<BTreeMap<String, AnalyzerSchema>> = OnceLock::new();
    REGISTRY.get_or_init(build_registry)
}

#[allow(clippy::too_many_lines)]
fn build_registry() -> BTreeMap<String, AnalyzerSchema> {
    let mut reg: BTreeMap<String, AnalyzerSchema> = BTreeMap::new();

    let mut schema = |id: &str, fields: &[(&str, FieldMeta)]| {
        let mut s = AnalyzerSchema::new();
        for (name, meta) in fields {
            s.insert((*name).to_string(), meta.clone());
        }
        reg.insert(id.to_string(), s);
    };

    schema(
        "static/complexity",
        &[
            ("function_complexity", FieldMeta::new("list", "function", "Per-function cyclomatic and cognitive complexity")),
            ("distribution", FieldMeta::new("aggregate", "", "Complexity distribution (simple/moderate/complex)")),
            ("high_risk_functions", FieldMeta::new("risk", "function", "Functions exceeding complexity thresholds")),
            ("aggregate", FieldMeta::new("aggregate", "", "Summary statistics")),
        ],
    );
    schema(
        "static/halstead",
        &[
            ("function_halstead", FieldMeta::new("list", "function", "Per-function Halstead volume, effort, and bugs")),
            ("distribution", FieldMeta::new("aggregate", "", "Effort distribution (low/medium/high/very_high)")),
            ("high_effort_functions", FieldMeta::new("risk", "function", "Functions with high Halstead effort")),
            ("aggregate", FieldMeta::new("aggregate", "", "Summary statistics")),
        ],
    );
    schema(
        "static/cohesion",
        &[
            ("function_cohesion", FieldMeta::new("list", "function", "Per-function LCOM cohesion score")),
            ("distribution", FieldMeta::new("aggregate", "", "Cohesion distribution")),
            ("low_cohesion_functions", FieldMeta::new("risk", "function", "Functions with poor cohesion")),
            ("aggregate", FieldMeta::new("aggregate", "", "Summary statistics")),
        ],
    );
    schema(
        "static/comments",
        &[
            ("comment_quality", FieldMeta::new("list", "comment", "Per-comment quality assessment")),
            ("function_documentation", FieldMeta::new("list", "function", "Per-function documentation status")),
            ("undocumented_functions", FieldMeta::new("risk", "function", "Functions lacking documentation")),
            ("aggregate", FieldMeta::new("aggregate", "", "Summary statistics")),
        ],
    );
    schema(
        "static/clones",
        &[
            ("clone_pairs", FieldMeta::new("list", "pair", "Detected clone pairs with similarity")),
            ("clone_type_distribution", FieldMeta::new("aggregate", "", "Clone type breakdown (Type-1/2/3)")),
            ("total_functions", FieldMeta::new("scalar", "", "Total functions analyzed")),
            ("total_clone_pairs", FieldMeta::new("scalar", "", "Total clone pairs (uncapped)")),
            ("clone_ratio", FieldMeta::new("scalar", "", "Fraction of functions involved in duplication")),
        ],
    );
    schema(
        "static/imports",
        &[
            ("import_list", FieldMeta::new("list", "import", "All import statements")),
            ("dependencies", FieldMeta::new("list", "dependency", "External dependencies with risk")),
            ("categories", FieldMeta::new("aggregate", "", "Import category breakdown")),
            ("aggregate", FieldMeta::new("aggregate", "", "Summary statistics")),
        ],
    );
    schema(
        "static/composition",
        &[
            ("breakdown", FieldMeta::new("aggregate", "", "File count per category")),
            ("percentages", FieldMeta::new("aggregate", "", "Percentage per category")),
            ("total_files", FieldMeta::new("scalar", "", "Total files analyzed")),
        ],
    );
    schema(
        "history/sentiment",
        &[
            ("time_series", FieldMeta::new("time_series", "tick", "Per-tick sentiment scores")),
            ("trend", FieldMeta::new("aggregate", "", "Sentiment trend direction")),
            ("low_sentiment_periods", FieldMeta::new("risk", "tick", "Ticks with negative sentiment")),
            ("aggregate", FieldMeta::new("aggregate", "", "Summary statistics")),
        ],
    );
    schema(
        "history/anomaly",
        &[
            ("anomalies", FieldMeta::new("list", "tick", "Detected anomalous ticks")),
            ("time_series", FieldMeta::new("time_series", "tick", "Per-tick anomaly metrics and z-scores")),
            ("aggregate", FieldMeta::new("aggregate", "", "Summary statistics")),
        ],
    );
    schema(
        "history/devs",
        &[
            ("developers", FieldMeta::new("list", "developer", "Per-developer contribution statistics")),
            ("languages", FieldMeta::new("list", "language", "Per-language contribution breakdown")),
            ("busfactor", FieldMeta::new("list", "language", "Bus factor per language")),
            ("activity", FieldMeta::new("time_series", "tick", "Per-tick commit activity by developer")),
            ("churn", FieldMeta::new("time_series", "tick", "Per-tick lines added/removed")),
            ("aggregate", FieldMeta::new("aggregate", "", "Summary statistics")),
        ],
    );
    schema(
        "history/file-history",
        &[
            ("file_churn", FieldMeta::new("list", "file", "Per-file change frequency and contributors")),
            ("file_contributors", FieldMeta::new("list", "file", "Per-file contributor breakdown")),
            ("hotspots", FieldMeta::new("risk", "file", "High-churn files")),
            ("composition", FieldMeta::new("aggregate", "", "File type composition")),
            ("composition_ts", FieldMeta::new("time_series", "tick", "File composition over time")),
            ("aggregate", FieldMeta::new("aggregate", "", "Summary statistics")),
        ],
    );
    schema(
        "history/couples",
        &[
            ("file_coupling", FieldMeta::new("list", "pair", "Co-changed file pairs")),
            ("developer_coupling", FieldMeta::new("list", "pair", "Developer collaboration pairs")),
            ("file_ownership", FieldMeta::new("list", "file", "Per-file ownership")),
            ("aggregate", FieldMeta::new("aggregate", "", "Summary statistics")),
        ],
    );
    schema(
        "history/shotness",
        &[
            ("node_hotness", FieldMeta::new("list", "node", "AST node change frequency")),
            ("node_coupling", FieldMeta::new("list", "pair", "Co-changed AST node pairs")),
            ("hotspot_nodes", FieldMeta::new("risk", "node", "Frequently changed nodes")),
            ("aggregate", FieldMeta::new("aggregate", "", "Summary statistics")),
        ],
    );
    schema(
        "history/burndown",
        &[
            ("global_survival", FieldMeta::new("time_series", "sample", "Global code survival curve")),
            ("file_survival", FieldMeta::new("list", "file", "Per-file survival data")),
            ("developer_survival", FieldMeta::new("list", "developer", "Per-developer survival data")),
            ("aggregate", FieldMeta::new("aggregate", "", "Summary statistics")),
        ],
    );
    schema(
        "history/quality",
        &[
            ("time_series", FieldMeta::new("time_series", "tick", "Per-tick code quality metrics")),
            ("aggregate", FieldMeta::new("aggregate", "", "Summary statistics")),
        ],
    );
    schema(
        "history/imports",
        &[
            ("import_list", FieldMeta::new("list", "import", "Import statements (requires UAST mode)")),
            ("dependencies", FieldMeta::new("list", "dependency", "Dependencies (requires UAST mode)")),
            ("categories", FieldMeta::new("aggregate", "", "Import category breakdown")),
            ("aggregate", FieldMeta::new("aggregate", "", "Summary statistics")),
        ],
    );
    schema(
        "history/typos",
        &[("typos", FieldMeta::new("list", "identifier", "Detected identifier typos (requires UAST mode)"))],
    );

    reg
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn known_analyzer_has_schema() {
        let s = schema_for_analyzer("static/complexity").expect("registered");
        let fc = s.get("function_complexity").expect("field");
        assert_eq!(fc.r#type, "list");
        assert_eq!(fc.grain, "function");
    }

    #[test]
    fn unknown_analyzer_yields_none() {
        assert!(schema_for_analyzer("history/nonexistent").is_none());
    }

    #[test]
    fn typos_has_single_field() {
        let s = schema_for_analyzer("history/typos").expect("registered");
        assert_eq!(s.len(), 1);
        assert!(s.contains_key("typos"));
    }

    #[test]
    fn registry_covers_all_go_entries() {
        // 17 analyzer IDs in schema_registry.go.
        for id in [
            "static/complexity",
            "static/halstead",
            "static/cohesion",
            "static/comments",
            "static/clones",
            "static/imports",
            "static/composition",
            "history/sentiment",
            "history/anomaly",
            "history/devs",
            "history/file-history",
            "history/couples",
            "history/shotness",
            "history/burndown",
            "history/quality",
            "history/imports",
            "history/typos",
        ] {
            assert!(schema_for_analyzer(id).is_some(), "missing {id}");
        }
    }
}
